### PR TITLE
Implementation of stateful retries for vulnerability analyzers

### LINKED ISSUE #124: Implement retry mechanism
As a majority of what the applications in this repo are doing consists of reaching out to external services, they are of course very vulnerable to instabilities in those external services affecting overall system performance and reliability.

There are various reasons for external requests going bad:

* Connection timeouts
* Rate limiting (HTTP 429)
* External system temporarily unavailable (HTTP 503)
* Many more

At the moment, whenever the streams applications encounter an error, the stream thread is shut down, and will not be available for further task processing (that's the default behavior in Kafka Streams). Not good for obvious reasons.

Simply ignoring such errors is also not a good option. 

According to some initial research, the idiomatic way of implementing retries in Kafka Streams is to have dedicated retry topics, and eventually a "dead letter topic" for when the maximum retries have exceeded. See https://www.confluent.io/blog/error-handling-patterns-in-kafka/

Having retries implemented in stream processors is most likely sub-optimal, as it will reduce message throughput of the consumer.

### PR DESCRIPTION
**Pull request by nscuro**
This PR introduces a stateful, non-blocking retry mechanism for vulnerability analyzers. The same mechanic can be used for all other areas that need to communicate with external services, for example the repository meta analysis.

---

Vulnerability analyzers have been refactored to [stream processors](https://kafka.apache.org/33/documentation/streams/developer-guide/processor-api#id2), which gives them direct access to Kafka Streams' state stores. State stores are backed by `changelog` topics in Kafka, making them [fault-tolerant](https://kafka.apache.org/33/documentation/streams/developer-guide/processor-api#id5). Within the application, the state stores are (currently) in-memory, but could easily be switched to persistent if need be.

Few technical notes on the implementation with processors:
* `Processor`s can only `forward` records to the topic they're connected to in the `Topology`, or drop records by simply not `forward`ing them. They can't publish records to arbitrary topics.
* Processors can't be CDI beans (e.g. `@ApplicationScoped`). This is because, when adding `Processor`s to the `Topology` a `ProcessorSupplier` must be provided. `ProcessorSupplier`s are [required to create new instances of `Processor` upon every invocation](https://kafka.apache.org/33/documentation/streams/developer-guide/processor-api#connecting-processors-and-state-stores). Kafka Streams will ensure that this happening at runtime, and fail if it isn't the case. This means we can't use `@Inject`, `@Transactional` and other declarative CDI-goodies in there, and need to fall back to programmatic APIs instead. In contrast to `Processor`s, `ProcessorSupplier`s **can** be `@ApplicationScoped` CDI beans.

This is how the relevant part of the topology looks now:

image

With analyzers now being `Processor`s, they can all maintain their own state stores for retries and batching (if supported). For batching, this allows us to accommodate for different batch sizes (e.g. Snyk will support different sizes than OSS Index).

Because there is a difference in how fast the analyzers can analyze components (batching will always be faster than individual analysis), separate topics for each analyzer have been introduced. This allows each analyzer to work on it's "queue" at its own pace, without blocking others. Additionally, it enables more fine-grained parallelism, for example by increasing partition count of the `-snyk` topic to 50, allowing up to 50 parallel requests to Snyk, while keeping the `-ossindex` topic at 3 partitions. See [Kafka Streams' parallelism model](https://docs.confluent.io/platform/current/streams/architecture.html#parallelism-model).

---

The retry logic is largely encapsulated in the abstract [`RetryingProcessor`](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/processor/retry/RetryingProcessor.java) class.

It converts incoming `Record`s into [`RetryableRecord`s](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/processor/retry/RetryableRecord.java), which include additional retry metadata like numbers of attempts, and the timestamp of the next attempt. 

Records that failed with a ["retryable reason"](https://github.com/mehab/DTKafkaPOC/blob/0b460d8d2e4596e47eb6569e470e5b711bec0ed4/vulnerability-analyzer/src/main/java/org/acme/processor/snyk/SnykProcessor.java#L126-L142) are written to a `-retry` state store using the [`scheduleForRetry`](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/processor/retry/RetryingProcessor.java#L122-L135) method. This method automatically detects when a record has exceeded the maximum amount of retries, and if so, invokes the [`onMaxRetriesExceeded`](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/processor/retry/RetryingProcessor.java#L115-L120) callback. Using [punctuators](https://kafka.apache.org/33/documentation/streams/developer-guide/processor-api#defining-a-stream-processor), the retry store is checked every second for records that are due for retry. For records that are due, the [`onRetry`](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/processor/retry/RetryingProcessor.java#L108-L113) callback is invoked. Delays between retries employ a [configurable](https://github.com/mehab/DTKafkaPOC/blob/ddbef5eca016f092b4c28967e4c818929bbaba45/vulnerability-analyzer/src/main/java/org/acme/config/RetryConfig.java) exponential back-off mechanism, with optional jitter. Retry behavior is configurable on a per-analyzer basis, e.g.:

https://github.com/mehab/DTKafkaPOC/blob/8b5ce80fbea9671df460f676040bba2b360f38c6/vulnerability-analyzer/src/main/resources/application.properties#L62-L66

Both the number of records present in the retry store, and the status of each processed record, are reported via Prometheus (`http://localhost:8092/q/metrics`):

image

---

[Circuit breakers](https://resilience4j.readme.io/docs/circuitbreaker) are added so that retries do not overwhelm the remote services. Per the [current default](https://resilience4j.readme.io/docs/circuitbreaker#create-and-configure-a-circuitbreaker), if 50% out of 100 requests to the external service fail, the circuit is opened for 60 seconds. Of course, circuit breakers too are configurable per-analyzer, e.g.:

https://github.com/mehab/DTKafkaPOC/blob/0b460d8d2e4596e47eb6569e470e5b711bec0ed4/vulnerability-analyzer/src/main/resources/application.properties#L46-L53

---

The `VulnerabilityResult` class has been extended with a `failureReason` field, that can be used to communicate *if* and *why* the analysis of a given record failed. Just a quick & dirty solution. 

---

The Snyk analyzer now supports multiple tokens, and will use them in a round-robin fashion. This resembles how we did it in DT 4.7.0.

---

Closes #124

### CODE REVIEW

**Comment by nscuro** [comment]
One major downside of this approach is that the natural back-pressure of Kafka consumers is kind of lost.

It will consume all messages immediately, put them into batches, and ultimately into the retry store. In other words, the state stores will grow quite large, especially if the endpoints are requiring frequent retries.

File: vulnerability-analyzer/src/main/java/org/acme/processor/snyk/SnykProcessor.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    protected void onMaxRetriesExceeded(final RetryableRecord<String, Component> record) {
+        reportFailure(record, new RuntimeException("Max retry attempts exceeded"));
+    }
+
+    private void analyze(final RetryableRecord<String, Component> record) {
+        final Page<Issue> issuePage;
+        try {
+            issuePage = cache.get(record.value().getPurl().getCoordinates(), purl -> {
```
[the remark this answers is not in the document]
Comment (nscuro) [reply vulnerability-analyzer/src/main/java/org/acme/processor/snyk/SnykProcessor.java:84]: @sonatype-lift ignoreall

**Comment by nscuro** [comment]
With every analyzer now having its own topic, I think there's little value in keeping the "per-identifier-type" topics. 

Untitled

Instead, analyzers could directly consume from `dtrack.vuln-analysis.component`, and re-key as needed. This would also make it easier to determine when an analyzer finished its work, as there wouldn't be multiple events for the same component anymore, if the component has more than one identifier.

image

Implementation-wise, it would look like this for the internal analyzer:

```java
if (internalScannerConfig.enabled()) {
    streamsBuilder
            .stream(KafkaTopic.VULN_ANALYSIS_COMPONENT.getName(), Consumed
                    .with(Serdes.UUID(), componentSerde)
                    .withName(processorNameConsume(KafkaTopic.VULN_ANALYSIS_COMPONENT)))
            .filter((uuid, component) -> component.getCpe() != null || component.getPurl() != null,
                    Named.as("filter_component_with_cpe_or_purl_for_internal_scanner"))
            .selectKey((uuid, component) -> {
                if (component.getPurl() != null) {
                    return component.getPurl().getCoordinates();
                } else {
                    return component.getCpe();
                }
            }, Named.as("re-key_to_cpe_or_purl_coordinates_for_internal_scanner"))
            .repartition(Repartitioned
                    .with(Serdes.String(), componentSerde)
                    .withName("internal-scanner"))
            .process(internalScannerProcessorSupplier, Named.as("analyze_cpe_or_purl_component_internally"))
            .to(KafkaTopic.VULN_ANALYSIS_VULNERABILITY.getName(), Produced
                    .with(Serdes.String(), vulnResultSerde)
                    .withName(processorNameProduce(KafkaTopic.VULN_ANALYSIS_VULNERABILITY, "internal_results")));
}
```

**Review by japurva1502 — APPROVED** [review]
LGTM

**Pull request merged**